### PR TITLE
Add CScriptNum decode python implementation in functional suite

### PR DESCRIPTION
**Pull request by instagibbs**
I needed this for reasons and thought it'd be good to upsteam it.

### CODE REVIEW

File: test/functional/test_framework/script.py [hunk 1]
```diff
@@ -371,11 +371,12 @@ def __init__(self, d=0):
 
     @staticmethod
     def encode(obj):
+        val = obj.value
```
Comment (promag) [inline test/functional/test_framework/script.py]: Could discard these unrelated these changes.

File: test/functional/test_framework/script.py [hunk 2]
```diff
@@ -385,6 +386,23 @@ def encode(obj):
...
 
+    @staticmethod
+    def decode(vch):
+        # We assume valid push_size and minimal encoding
+        value = vch[1:]
+        # Mask for all but the highest result bit
+        num_mask = (2**(len(value)*8) - 1) >> 1
+        if len(value) == 0:
+            return 0
+        else :
```
Comment (promag) [inline test/functional/test_framework/script.py]: nit, remove extra space.

**Review by promag — COMMENTED** [review]
Restarted appveyor.

> I needed this for reasons and thought it'd be good to upsteam it.

Could add one reason? otherwise this is dead code.

File: test/functional/test_framework/script.py (hunk 1 above)
Comment (instagibbs) [reply test/functional/test_framework/script.py]: it reduces the verbosity of it, I can remove but seemed better to me.

**Comment by instagibbs** [comment]
@promag it can(should?) be used in tests in the future. Allows extraction of blockheight from the coinbase transaction, for example.

File: test/functional/test_framework/script.py (hunk 2 above)
Comment (instagibbs) [reply test/functional/test_framework/script.py]: done

**Comment by instagibbs** [comment]
Addressed issues, and added a single usage in a test.

File: test/functional/test_framework/script.py [hunk 3]
```diff
@@ -385,6 +385,23 @@ def encode(obj):
             r[-1] |= 0x80
         return bytes([len(r)]) + r
 
+    @staticmethod
+    def decode(vch):
+        # We assume valid push_size and minimal encoding
+        value = vch[1:]
+        # Mask for all but the highest result bit
+        num_mask = (2**(len(value)*8) - 1) >> 1
```
Comment (promag) [inline test/functional/test_framework/script.py]: nit, could move this to L401.

File: test/functional/test_framework/script.py [hunk 4]
```diff
@@ -385,6 +385,23 @@ def encode(obj):
...
+        if len(value) == 0:
+            return 0
+        else:
```
Comment (promag) [inline test/functional/test_framework/script.py]: nit, could remove else branch, and above just `return 0 if len(value) == 0`.

**Review by promag — COMMENTED** [review]
Thanks for adding the test, I think it makes this PR more complete. I'm leaving two nits for your consideration.

utACK 9f9bd43.

File: test/functional/test_framework/script.py [hunk 5]
```diff
@@ -385,6 +385,23 @@ def encode(obj):
...
+            result = 0
+            for i in range(len(value)):
```
Comment (practicalswift) [inline test/functional/test_framework/script.py]: If @promag:s suggestion is not implemented: use `enumerate` here :-)

File: test/functional/test_framework/script.py [hunk 6]
```diff
@@ -385,6 +385,23 @@ def encode(obj):
             r[-1] |= 0x80
         return bytes([len(r)]) + r
 
+    @staticmethod
+    def decode(vch):
+        # We assume valid push_size and minimal encoding
+        value = vch[1:]
+        # Mask for all but the highest result bit
+        num_mask = (2**(len(value)*8) - 1) >> 1
+        if len(value) == 0:
```
Comment (practicalswift) [inline test/functional/test_framework/script.py]: If @promag:s suggestions is not implemented: Use `if not value:`

Comment (instagibbs) [reply test/functional/test_framework/script.py]: fwiw I wouldn't know hot to read that suggestion properly, sorry

File: test/functional/test_framework/script.py (hunk 1 above)
Comment (promag) [reply test/functional/test_framework/script.py]: I understand but IMO it's better without.

Comment (instagibbs) [reply test/functional/test_framework/script.py]: already removed

Comment (promag) [reply test/functional/test_framework/script.py]: Yeah I know :)

File: test/functional/mining_basic.py [hunk 7]
```diff
@@ -65,11 +65,15 @@ def assert_submitblock(block, result_str_1, result_str_2=None):
...
         assert 'coinbasetxn' not in tmpl
 
-        coinbase_tx = create_coinbase(height=int(tmpl["height"]) + 1)
+        next_height = int(tmpl["height"])
+        coinbase_tx = create_coinbase(height=next_height)
         # sequence numbers must not be max for nLockTime to have effect
         coinbase_tx.vin[0].nSequence = 2 ** 32 - 2
         coinbase_tx.rehash()
 
+        # round-tip the encoded bip34 block height commitment
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/mining_basic.py]: fixed

**Review by scravy — APPROVED** [review]
utACK

**Comment by promag** [comment]
utACK 2012d4d, nice to have more usages.

**Review by maguayo — APPROVED** [review]
utACK
